File: src/main/java/com/xpdustry/nohorny/NoHornyPlugin.java
// SPDX-License-Identifier: MIT
package com.xpdustry.nohorny;

import arc.ApplicationListener;
import arc.Core;
import com.github.mizosoft.methanol.Methanol;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.Strictness;
import com.xpdustry.nohorny.authentication.AllowAllAuthenticator;
import com.xpdustry.nohorny.authentication.ApiKeyAuthenticator;
import com.xpdustry.nohorny.authentication.Authenticator;
import com.xpdustry.nohorny.authentication.AuthenticatorConfig;
import com.xpdustry.nohorny.authentication.LocalhostAuthenticator;
import com.xpdustry.nohorny.authentication.MindustryServerAuthenticator;
import com.xpdustry.nohorny.classification.Classifier;
import com.xpdustry.nohorny.classification.ClassifierConfig;
import com.xpdustry.nohorny.classification.SightEngineClassifier;
import com.xpdustry.nohorny.classification.ViTClassifier;
import com.xpdustry.nohorny.config.FriendlyDurationDecoder;
import com.xpdustry.nohorny.config.InetAddressDecoder;
import com.xpdustry.nohorny.config.NoHornyConfig;
import com.xpdustry.nohorny.config.SealedConfigDecoder;
import com.xpdustry.nohorny.config.SnakeYamlLoader;
import com.xpdustry.nohorny.geometry.ImmutablePoint2;
import com.xpdustry.nohorny.image.ImageBinaryCodec;
import com.xpdustry.nohorny.image.ImageRenderer;
import com.xpdustry.nohorny.tracking.AutoModerator;
import com.xpdustry.nohorny.tracking.CanvasTracker;
import com.xpdustry.nohorny.tracking.DisplayTracker;
import java.net.http.HttpClient;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import mindustry.Vars;
import mindustry.mod.Plugin;
import org.github.gestalt.config.builder.GestaltBuilder;
import org.github.gestalt.config.decoder.BooleanDecoder;
import org.github.gestalt.config.decoder.DoubleDecoder;
import org.github.gestalt.config.decoder.EnumDecoder;
import org.github.gestalt.config.decoder.FloatDecoder;
import org.github.gestalt.config.decoder.IntegerDecoder;
import org.github.gestalt.config.decoder.ListDecoder;
import org.github.gestalt.config.decoder.OptionalDecoder;
import org.github.gestalt.config.decoder.RecordDecoder;
import org.github.gestalt.config.decoder.StringDecoder;
import org.github.gestalt.config.decoder.URIDecoder;
import org.github.gestalt.config.path.mapper.KebabCasePathMapper;
import org.github.gestalt.config.source.ClassPathConfigSourceBuilder;
import org.github.gestalt.config.source.FileConfigSourceBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class NoHornyPlugin extends Plugin {

    private static final Logger logger = LoggerFactory.getLogger(NoHornyPlugin.class);

    private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(
            Runtime.getRuntime().availableProcessors() * 2,
            Thread.ofVirtual().name("nohorny-worker-thread-", 0).factory());
    private final Gson gson = new GsonBuilder()
            .registerTypeAdapter(ImmutablePoint2.class, new ImmutablePoint2.TypeAdapter())
            .setStrictness(Strictness.STRICT)
            .create();
    private final Methanol http = Methanol.newBuilder()
            .followRedirects(HttpClient.Redirect.ALWAYS)
            .executor(this.executor)
            .build();
    private final Path directory = Vars.mods.getConfigFolder(this).file().toPath();
    private final List<NoHornyListener> listeners = new ArrayList<>();

    @Override
    public void init() {
        try {
            Files.createDirectories(this.directory);

            final var config = this.loadConfig();
            final var codec = new ImageBinaryCodec();
            final var renderer = new ImageRenderer();

            if (config.server().isPresent()) {
                final var server = config.server().orElseThrow();
                final var classifiers = server.classifiers().stream()
                        .map(this::createClassifier)
                        .toList();
                this.listeners.addAll(classifiers);
                final var authenticators = server.authenticators().stream()
                        .map(this::createAuthenticator)
                        .toList();
                this.listeners.addAll(authenticators);
                this.listeners.add(new NoHornyServer(
                        server, this.executor, this.gson, classifiers, authenticators, renderer, codec));
            }

            if (config.client().isPresent()) {
                final var clientConfig = config.client().orElseThrow();
                final var client = new NoHornyClient(this.executor, this.http, this.gson, clientConfig, codec);
                this.listeners.add(client);
                this.listeners.add(new AutoModerator(clientConfig.autoMod()));
                this.listeners.add(new DisplayTracker(this.executor, client, clientConfig.displays()));
                this.listeners.add(new CanvasTracker(this.executor, client, clientConfig.canvases()));
            }

            int initializedCount = 0;
            try {
                for (; initializedCount < this.listeners.size(); initializedCount++) {
                    this.listeners.get(initializedCount).onInit();
                }
            } catch (final Exception e) {
                this.closeListeners(this.listeners, initializedCount - 1);
                NoHornyPlugin.this.executor.shutdown();
                throw e;
            }
        } catch (final Exception e) {
            throw new RuntimeException("Failed to initialize nohorny", e);
        }

        logger.info("NoHorny plugin initialized");

        Core.app.addListener(new ApplicationListener() {

            @Override
            public void dispose() {
                closeListeners(NoHornyPlugin.this.listeners, NoHornyPlugin.this.listeners.size() - 1);
                NoHornyPlugin.this.http.shutdown();
                NoHornyPlugin.this.executor.shutdown();
            }
        });
    }

    private void closeListeners(final List<NoHornyListener> listeners, final int lastIndex) {
        for (int i = lastIndex; i >= 0; i--) {
            try {
                listeners.get(i).onExit();
            } catch (final Throwable e) {
                logger.error("Failed to close {}", listeners.get(i), e);
            }
        }
    }

    private Classifier createClassifier(final ClassifierConfig config) {
        return switch (config) {
            case ClassifierConfig.ViT vit ->
                new ViTClassifier(vit.labels(), vit.nsfwLabel(), this.directory.resolve(vit.file()), vit.thresholds());
            case ClassifierConfig.SightEngine se ->
                new SightEngineClassifier(this.http, this.gson, this.executor, se.user(), se.secret(), se.thresholds());
        };
    }

    private Authenticator createAuthenticator(final AuthenticatorConfig config) {
        return switch (config) {
            case AuthenticatorConfig.AllowAll ignored -> new AllowAllAuthenticator();
            case AuthenticatorConfig.Localhost ignored -> new LocalhostAuthenticator();
            case AuthenticatorConfig.ApiKey ignored ->
                new ApiKeyAuthenticator(this.directory.resolve("api-keys.json"), this.gson);
            case AuthenticatorConfig.MindustryServerList msl ->
                new MindustryServerAuthenticator(
                        this.http, this.gson, msl.sources(), msl.refreshInterval(), this.executor);
        };
    }

    private NoHornyConfig loadConfig() throws Exception {
        final var builder = new GestaltBuilder()
                .addSource(ClassPathConfigSourceBuilder.builder()
                        .setResource("com/xpdustry/nohorny/default-config.yaml")
                        .build())
                .setTreatMissingDiscretionaryValuesAsErrors(false)
                .addPathMapper(new KebabCasePathMapper())
                .addConfigLoader(new SnakeYamlLoader())
                .addDecoder(new RecordDecoder())
                .addDecoder(new URIDecoder())
                .addDecoder(new BooleanDecoder())
                .addDecoder(new DoubleDecoder())
                .addDecoder(new FriendlyDurationDecoder())
                .addDecoder(new EnumDecoder<>())
                .addDecoder(new FloatDecoder())
                .addDecoder(new IntegerDecoder())
                .addDecoder(new OptionalDecoder())
                .addDecoder(new StringDecoder())
                .addDecoder(new SealedConfigDecoder())
                .addDecoder(new ListDecoder())
                .addDecoder(new InetAddressDecoder());
        final var file = this.directory.resolve("config.yaml");
        if (Files.exists(file)) {
            builder.addSource(FileConfigSourceBuilder.builder().setPath(file).build());
        }
        final var gestalt = builder.build();
        gestalt.loadConfigs();
        return gestalt.getConfig("", NoHornyConfig.class);
    }
}

File: src/main/kotlin/com/xpdustry/nohorny/NoHornyPlugin.kt
/*
 * This file is part of NoHorny. The plugin securing your server against nsfw builds.
 *
 * MIT License
 *
 * Copyright (c) 2024 Xpdustry
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.xpdustry.nohorny

import arc.ApplicationListener
import arc.Core
import arc.util.CommandHandler
import arc.util.OS
import com.sksamuel.hoplite.ConfigException
import com.sksamuel.hoplite.ConfigLoaderBuilder
import com.sksamuel.hoplite.KebabCaseParamMapper
import com.sksamuel.hoplite.addPathSource
import com.xpdustry.nohorny.image.ImageProcessorImpl
import com.xpdustry.nohorny.image.ImageRendererImpl
import com.xpdustry.nohorny.image.analyzer.AnalyzerConfig
import com.xpdustry.nohorny.image.analyzer.DebugImageAnalyzer
import com.xpdustry.nohorny.image.analyzer.FallbackAnalyzer
import com.xpdustry.nohorny.image.analyzer.ImageAnalyzer
import com.xpdustry.nohorny.image.analyzer.SightEngineAnalyzer
import com.xpdustry.nohorny.image.cache.H2ImageCache
import com.xpdustry.nohorny.image.cache.ImageCache
import com.xpdustry.nohorny.image.cache.ImageCacheConfig
import com.xpdustry.nohorny.tracker.CanvasesTracker
import com.xpdustry.nohorny.tracker.DisplaysTracker
import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import mindustry.Vars
import mindustry.mod.Plugin
import okhttp3.Dispatcher
import okhttp3.OkHttpClient
import org.slf4j.LoggerFactory
import java.lang.Runnable
import java.nio.file.Files
import java.util.concurrent.Executors
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger
import kotlin.io.path.absolutePathString
import kotlin.io.path.notExists
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toJavaDuration

public class NoHornyPlugin : Plugin() {
    private val directory = Vars.modDirectory.child("nohorny").file().toPath()
    private val file = directory.resolve("config.yaml")
    private val listeners = mutableListOf<NoHornyListener>()
    private val executor = Executors.newScheduledThreadPool(4, NoHornyThreadFactory)
    private val logger = LoggerFactory.getLogger(NoHornyPlugin::class.java)

    private val loader =
        ConfigLoaderBuilder.empty()
            .withClassLoader(javaClass.classLoader)
            .addDefaultDecoders()
            .addDefaultParamMappers()
            .addDefaultParsers()
            .addParameterMapper(KebabCaseParamMapper)
            .addPathSource(file)
            .withReport()
            .withReportPrintFn(logger::debug)
            .strict()
            .build()

    private val http =
        OkHttpClient.Builder()
            .connectTimeout(10.seconds.toJavaDuration())
            .connectTimeout(10.seconds.toJavaDuration())
            .readTimeout(10.seconds.toJavaDuration())
            .writeTimeout(10.seconds.toJavaDuration())
            .dispatcher(Dispatcher(Executors.newScheduledThreadPool(4, NoHornyThreadFactory)))
            .build()

    internal var config = NoHornyConfig()
    private var analyzer: ImageAnalyzer = ImageAnalyzer.None

    init {
        Files.createDirectories(directory)
    }

    override fun init() {
        reload()
        val renderer = ImageRendererImpl
        val processor = ImageProcessorImpl({ analyzer }, createCache(config.imageCache), renderer)
        listeners += CanvasesTracker({ config }, processor)
        listeners += DisplaysTracker({ config }, processor)
        listeners += NoHornyAutoMod(this)

        listeners.forEach(NoHornyListener::onInit)
        addExitListener { executor.shutdown() }
        logger.info("Initialized no-horny, to the horny jail we go.")
    }

    override fun registerServerCommands(handler: CommandHandler) {
        handler.register<Unit>("nohorny-reload", "Reload nohorny config.") { _, _ ->
            try {
                reload()
                logger.info("Reloaded config")
            } catch (e: ConfigException) {
                logger.error(e.message!!)
            } catch (e: Exception) {
                logger.error("Failed to reload config", e)
            }
        }
    }

    private fun reload() {
        if (file.notExists()) {
            analyzer = ImageAnalyzer.None
            return
        }
        val config = loader.loadConfigOrThrow<NoHornyConfig>()
        this.config = config
        this.analyzer = createAnalyzer(config.analyzer)
    }

    private fun createAnalyzer(config: AnalyzerConfig): ImageAnalyzer =
        when (config) {
            is AnalyzerConfig.None -> ImageAnalyzer.None
            is AnalyzerConfig.Debug -> DebugImageAnalyzer(directory.resolve("debug"))
            is AnalyzerConfig.Fallback -> FallbackAnalyzer(createAnalyzer(config.primary), createAnalyzer(config.secondary))
            is AnalyzerConfig.SightEngine -> SightEngineAnalyzer(config, http)
        }

    private fun createCache(config: ImageCacheConfig): ImageCache =
        when (config) {
            is ImageCacheConfig.None -> ImageCache.None
            is ImageCacheConfig.Local -> {
                val hikari =
                    HikariDataSource(
                        HikariConfig().apply {
                            jdbcUrl = "jdbc:h2:${directory.resolve("database.h2").absolutePathString()};MODE=MYSQL"
                            poolName = "no-horny-jdbc-pool"
                            maximumPoolSize = OS.cores * 2
                            minimumIdle = 1
                        },
                    )
                addExitListener { hikari.close() }
                H2ImageCache(hikari, config)
            }
        }

    private fun addExitListener(block: () -> Unit) {
        Core.app.listeners.add(
            object : ApplicationListener {
                override fun dispose() = block()

                override fun toString() = "NoHornyExitListener"
            },
        )
    }

    private object NoHornyThreadFactory : ThreadFactory {
        private val count = AtomicInteger(0)

        override fun newThread(runnable: Runnable) =
            Thread(runnable, "nohorny-worker-${count.incrementAndGet()}")
                .apply { isDaemon = true }
    }
}
